src/NServiceBus.Core/Transports/DeliveryConstraintCollection.cs

```diff
@@ -0,0 +1,44 @@
+namespace NServiceBus.Transports
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using NServiceBus.DeliveryConstraints;
+
+    /// <summary>
+    /// A collection of delivery constraints.
+    /// </summary>
+    public class DeliveryConstraintCollection
+    {
+        List<DeliveryConstraint> constraints;
+
+        internal DeliveryConstraintCollection(IEnumerable<DeliveryConstraint> constraints)
+        {
+            this.constraints = constraints.ToList();
+        }
+
+        /// <summary>
+        /// Tries to get and remove constraint of a given type.
+        /// </summary>
+        /// <typeparam name="T">Requested type.</typeparam>
+        /// <param name="constraint">Returned constraint or null.</param>
+        /// <returns>Success or failure</returns>
+        public bool TryRemove<T>(out T constraint) where T : DeliveryConstraint
```

Comment:
@timbussmann can u point me to the discussion on out parameters. it is a fairly well accepted pattern in .net. and yes Maybe is an understood concept my people who are watching the f# community. but i am not sure that is a large percentage